### DISCUSSION TITLE
Serving files directly from S3 provider

### DISCUSSION THREAD

**Discussion by CovertCode** [pocketbase/pocketbase#4046] [Q&A]
I linked my backblaze S3 account to pockethost and created a record. But the file(image in my case) was served from pockethost instance not directly from S3. Although while file is created in S3 all those folders are also created in S3. Won't it be possible for pockethost to just store files path and append it to S3 endpoint?
Let's say this is a file which was uploaded by pocketbase: `https://pockethost.s3.us-west-002.backblazeb2.com/wzhkbojlwpp25e9/gxma9jbxnh8iajh/moon_profile_halftone_pattern_editorial_illust_uKkGl4knKL.jpg`
When user enters the S3 credentials he can also enter the public URL and to which file path made by pockethost can be appended then store it in file url field in the collection.

Because if file is served always from pockethost then it will create a bandwidth issue and won't be able to utilize S3 providers speed.

**Comment by ganigeorgiev** [comment]
All S3 operations are performed over the S3 APIs. Note that the bucket can be also private and not always public. If your bucket is public you are free to access the file with the same prefix via your CDN/S3 public endpoint, aka. `https://yourcdn.com/:collectionId/:recordId/:filename`.
The only drawback is that thumbs will not work since they are generated lazily on first access by the PocketBase server.

See also https://github.com/pocketbase/pocketbase/discussions/2191, https://github.com/pocketbase/pocketbase/issues/28#issuecomment-1492516269.

**Comment by CovertCode** [comment]
I think it would be better to handle images by myself. As I will also have to convert images to webp then compress it with specific library. Storing image/file urls in database would be a better and more performant option.

**Reply by benallfree** [reply]
Just to add to this... rather than storing the URLs, see if you can make them some kind of deterministic pattern that could be derived from data you already have when the page renders.

**Reply by CovertCode** [reply]
If I perform image processing first and then upload the file via Pocketbase, then it will work. As @ganigeorgiev said [**https://yourcdn.com/:collectionId/:recordId/:filename**](https://yourcdn.com/:collectionId/:recordId/:filename), this URL pattern will work for everyone. The only additional thing I would have to do is create a collection that will store my S3 URL.

@ganigeorgiev, is it possible to add a public URL of the bucket in S3 settings and get it via SDK (`pb_hooks`)?

Backblaze gives a public URL like: [**https://f002.backblazeb2.com/file/pockethost/wzhkbojlwpp25e9/gxma9jbxnh8iajh/moon_profile_halftone_pattern_editorial_illust_uKkGl4knKL.jpg**](https://f002.backblazeb2.com/file/pockethost/wzhkbojlwpp25e9/gxma9jbxnh8iajh/moon_profile_halftone_pattern_editorial_illust_uKkGl4knKL.jpg)

So, users can just store [**https://f002.backblazeb2.com_/file/pockethost/**](https://f002.backblazeb2.com_/file/pockethost/) in settings and retrieve it instantly, which would do the work. Then, S3-friendly URLs will be generated easily. Otherwise, public URLs need to be stored in a collection or hardcoded.

**Reply by benallfree** [reply]
I think Cloudflare may also have an on-demand image conversion service. Sorry I don't know much more about it!
